core/controllers/question_player.py

```diff
@@ -0,0 +1,73 @@
+# Copyright 2019 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Controllers for the questions player."""
+
+from core.controllers import acl_decorators
+from core.controllers import base
+from core.domain import skill_domain
+from core.domain import skill_services
+import utils
+
+
+class SkillMasteryDataHandler(base.BaseHandler):
+    """A handler that handles fetching and updating the degrees of user
+    skill mastery.
+    """
+
+    @acl_decorators.can_access_user_skill_mastery
+    def put(self):
+        """Handles PUT requests."""
+        degree_of_mastery_per_skill = (
```

Comment:
Just one question -- any reason why we are just trusting this info from the frontend, rather than taking the pass/fail info and (re-)calculating the updated skill mastery in the backend?

It seems like the former would be more fragile since you lose all the info about what the learner actually did that caused this particular update. So if we were to change how we score things in the future, we wouldn't have any information to go on in order to apply that new scoring mechanism to existing data...

Reply:
Skill mastery is not only calculated based on pass/fail, but also calculated based on the number of wrong answers, hint viewed, and solution viewed. It would be too much to store that information and I don't really see much purpose of doing that, since the information would not be reused unless the scoring mechanism is changed. Even though it will be changed in the future, we don't have to change the existing data.

Reply:
OK, I'm fine with that, though can you explain your last sentence a bit more? I am not sure I understand why we wouldn't need to change the existing data. What is the alternative?

Reply:
I think it would be okay to keep the existing skill mastery degree, and increase or decrease based on that.